### PR TITLE
[prisma][node.js][react] Finished implementation of recommendation by adding more variables into account and adding the display in the front end with adding functionality.

### PR DESCRIPTION
**Pull request by DiegoLandaeta03**
**Prisma**

- Changed the way that the stitch is created in the backend so that mood, dance, and mix values are percentages that add up to 1. 

**Backend**

- Added tonality scoring that takes a look at the last song on the stitch's key and mode and determines if the transition is good with a recommended song based on the Camelot System. 
- Also implemented the weighing of different variables for mood and mixability. 

**Frontend** 

- Created fetching for recommended songs that begins once the user has at least 1 song added in their stitch. 
- When stitch is empty nothing is fetched. 
- Recommended list changes every time the user adds or deletes a song from their stitch.

**Image of how recommended looks, I sent too many requests so now I have to wait so that is why I will not be sending a video. Glad I took this picture before API stopped working.**

Screenshot 2024-07-17 at 3 20 11 PM

### CODE REVIEW

File: back-end/routes/recommendations.js [hunk 1]
```diff
@@ -63,59 +63,114 @@ function calculateGlobalFeatureAverages(features) {
...
     const lastSongFeatures = features.length > 0 ? features[features.length - 1] : null;
     return {
-        averages,
+        stitchAverages,
         lastSongFeatures
     };
 }
 
-function gradingAlgorithm(lastSongFeatures, recommendationAudioFeatures, averages) {
+function tonalityScore(lastSongFeatures, currentSongFeatures) {
```
Comment (alegleason) [inline back-end/routes/recommendations.js:76]: nit

`function getTonalityScore(lastSongFeatures, currentSongFeatures) {`

File: back-end/routes/recommendations.js [hunk 2]
```diff
@@ -63,59 +63,114 @@ function calculateGlobalFeatureAverages(features) {
...
+        return 0;
+    }
+
+    // Moving either to inside or outside circle on Camelot System (ex: A-flat major to F-minor)
+    if (lastSongMode == 0) {
+        if (currentSongMode == 1 && currentSongKey == (lastSongKey + 3) % 12) {
+            return 1;
+        }
+    }
+    else {
```
Comment (alegleason) [inline back-end/routes/recommendations.js:101]: nit 

`} else {`

As that's how I've seen it's been used throughout the app

File: back-end/routes/recommendations.js [hunk 3]
```diff
@@ -63,59 +63,114 @@ function calculateGlobalFeatureAverages(features) {
...
+        const response = await fetch(`https://api.spotify.com/v1/tracks?ids=${commaSeparatedIds}`, {
+            headers: {
+                'Authorization': `Bearer ${accessToken}`
+            }
+        });
+        const trackData = await response.json();
+        const tracks = trackData.tracks;
+        return tracks;
+    } catch (error) {
+        console.error(error);
```
Comment (alegleason) [inline back-end/routes/recommendations.js:170]: Please do not commit `console.error`s

File: back-end/routes/stitchRoutes.js [hunk 4]
```diff
@@ -56,22 +56,27 @@ router.get('/title/:stitchId', async (req, res) => {
...
     const user = await prisma.user.findUnique({
         where: {
             username: username
         }
     })
 
+    const totalPreferences = mood + dance + mix;
+    const moodWeight = mood / totalPreferences;
+    const danceWeight = dance / totalPreferences;
+    const mixWeight = 1 - (moodWeight + danceWeight);
```
Comment (alegleason) [inline back-end/routes/stitchRoutes.js:70]: Any reason why we're just not doing `const mixWeight  =  mix / totalPreferences;` ?

File: back-end/routes/stitchRoutes.js [hunk 5]
```diff
@@ -56,22 +56,27 @@ router.get('/title/:stitchId', async (req, res) => {
...
-    const { title, username, mood, dance, mix, explore } = req.body
+    const { title, username, mood, dance, mix, explore, imageUrl } = req.body
 
     const user = await prisma.user.findUnique({
         where: {
             username: username
         }
     })
 
+    const totalPreferences = mood + dance + mix;
```
Comment (alegleason) [inline back-end/routes/stitchRoutes.js:67]: What happens if the user has zero for all values? Which according to the UI, is possible.

File: back-end/routes/stitchRoutes.js (hunk 4 above)
Comment (DiegoLandaeta03) [reply back-end/routes/stitchRoutes.js:70]: > Any reason why we're just not doing `const mixWeight = mix / totalPreferences;` ?

I was doing that to make sure that the weights add up to 1.

File: back-end/routes/stitchRoutes.js (hunk 5 above)
Comment (DiegoLandaeta03) [reply back-end/routes/stitchRoutes.js:67]: Good point, resolved by checking if totalPreferences is equal to 0. If true I set all weights to 0, and then for recommendations I am returning the normal Spotify recommendations now if all are 0.

File: back-end/routes/recommendations.js (hunk 3 above)
Comment (roeshanigos) [reply back-end/routes/recommendations.js:170]: best practice is- use a logger to log your errors so you have a paper trail of errors that have occurred in your app

Comment (DiegoLandaeta03) [reply back-end/routes/recommendations.js:170]: Awesome, let's discuss this in our 1:1 tomorrow!

File: back-end/routes/stitchRoutes.js (hunk 4 above)
Comment (alegleason) [reply back-end/routes/stitchRoutes.js:70]: Yeah that's okay

**Pull request merged**